lock_api/src/mutex.rs

```diff
@@ -591,79 +597,93 @@ unsafe impl<'a, R: RawMutex + 'a, T: ?Sized + 'a> StableAddress for MutexGuard<'
 /// An RAII mutex guard returned by `MutexGuard::map`, which can point to a
 /// subfield of the protected data.
 ///
-/// The main difference between `MappedMutexGuard` and `MutexGuard` is that the
+/// The main difference between `OwnedMappedMutexGuard` and `MutexGuard` is that the
 /// former doesn't support temporarily unlocking and re-locking, since that
 /// could introduce soundness issues if the locked object is modified by another
 /// thread.
 #[must_use = "if unused the Mutex will immediately unlock"]
-pub struct MappedMutexGuard<'a, R: RawMutex, T: ?Sized> {
+pub struct OwnedMappedMutexGuard<'a, R: RawMutex, T: 'a> {
     raw: &'a R,
-    data: *mut T,
-    marker: PhantomData<&'a mut T>,
+    // We use `&'a mut` to make this type invariant over `'a`
+    marker: PhantomData<&'a mut ()>,
+    data: T,
 }
 
-unsafe impl<'a, R: RawMutex + Sync + 'a, T: ?Sized + Sync + 'a> Sync
-    for MappedMutexGuard<'a, R, T>
-{
-}
-unsafe impl<'a, R: RawMutex + 'a, T: ?Sized + Send + 'a> Send for MappedMutexGuard<'a, R, T> where
+/// An `OwnedMappedMutexGuard` that contains a mutable pointer (for backwards compatibility)
+pub type MappedMutexGuard<'a, R, T> = OwnedMappedMutexGuard<'a, R, &'a mut T>;
+
+unsafe impl<'a, R: RawMutex + Sync + 'a, T: Sync + 'a> Sync for OwnedMappedMutexGuard<'a, R, T> {}
+unsafe impl<'a, R: RawMutex + 'a, T: Send + 'a> Send for OwnedMappedMutexGuard<'a, R, T> where
     R::GuardMarker: Send
 {
 }
 
-impl<'a, R: RawMutex + 'a, T: ?Sized + 'a> MappedMutexGuard<'a, R, T> {
-    /// Makes a new `MappedMutexGuard` for a component of the locked data.
+impl<'a, R: RawMutex + 'a, T: 'a> OwnedMappedMutexGuard<'a, R, T> {
+    /// Makes a new `OwnedMappedMutexGuard` for a component of the locked data.
     ///
-    /// This operation cannot fail as the `MappedMutexGuard` passed
+    /// This operation cannot fail as the `OwnedMappedMutexGuard` passed
     /// in already locked the mutex.
     ///
     /// This is an associated function that needs to be
-    /// used as `MappedMutexGuard::map(...)`. A method would interfere with methods of
+    /// used as `OwnedMappedMutexGuard::map(...)`. A method would interfere with methods of
     /// the same name on the contents of the locked data.
     #[inline]
-    pub fn map<U: ?Sized, F>(s: Self, f: F) -> MappedMutexGuard<'a, R, U>
+    pub fn map<U: 'a, F>(s: Self, f: F) -> OwnedMappedMutexGuard<'a, R, U>
     where
-        F: FnOnce(&mut T) -> &mut U,
+        F: FnOnce(T) -> U,
     {
+        use core::ptr;
+
+        let mut s = mem::ManuallyDrop::new(s);
+
         let raw = s.raw;
-        let data = f(unsafe { &mut *s.data });
-        mem::forget(s);
-        MappedMutexGuard {
+
+        let data = unsafe { ptr::read(&mut s.data) };
+
+        let data = f(data);
```

Comment:
Same

Reply:
There is a potential double-drop here: if `f` panics then `data` will be dropped again despite having already been read from `s.data`.